### ISSUE TITLE
UI scaling wider than intended when Incognito button is enabled

### ISSUE THREAD

**Issue by pressRtowin** [TrianguloY/UrlChecker#230]
### Steps to reproduce

1. Set incognito to anything other than always on/off
2. Launch URLCheck on a large screen or low dpi (high dp/minimum width) device
3. Observe chonky boi

### Expected behavior

UI scaling is not affected by the presence of the Incognito button

### Actual behavior

UI scaling is affected by the presence of the Incognito button

### URLCheck version

Uhh the dev version after 2.11 that fixes legacy shortcuts 😅

### Android version

13

### Android Custom/Specific ROM or Device

One UI 5.1

### Other details

> The only difference is the incognito button. What happens if you hide it?

Well, I found out. 😅

> I'll check the layout again just in case, but maybe I should add a maximum width? Hmm.

I honestly don't mind the wider scaling aesthetically. As I said in the other thread:

> While overall it does feel more in place on a larger device now, some modules start to suffer from uh . . . "Too Wide Syndrome"?

If the widenees were made to be intentional, it could work fine with some small QoL UI adjustments to some modules. I'm thinking something along the lines of alternate row shading,  lines, or something else functionally equivalent that bridges the left and right extremes.

> Can you create a new issue for this? It will be easier to manage.

Done. 😛

Pasting the screenshots here for the record:

Old version:
![Screenshot_20230514_044657_URLCheck](https://github.com/TrianguloY/UrlChecker/assets/25999169/5afaf747-204a-4f5c-8bd9-46cdf75bca86)

New version:
![Screenshot_20230514_044734_URLCheck](https://github.com/TrianguloY/UrlChecker/assets/25999169/2f14276c-c75b-4f35-b5c7-4db0d98e47e6)

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I will fill out all of the requested information in this form.

**Comment by TrianguloY** [comment]
Perfect, thanks!

As I said in #228 a solution would be to add a maxWidth setting, but the idea to adapt modules to work on wider screens should be considered too.

Also, I guess the issue will also happen with the custom tabs button, and maybe with other modules configurations.

**Comment by TrianguloY** [comment]
> > As I said in #228 a solution would be to add a maxWidth setting, but the idea to adapt modules to work on wider screens should be considered too.
> 
> The current (intended) UI doesn't look _bad_ on larger screens by any means. But I'm always for better tablet UI optimizations. Maybe there's other ways URLCheck could take advantage of the added real estate? A multi-column layout perhaps? 

Multi column is definitely something I wanted to add, even for small screens there are some modules that are quite small, so it could benefit from it, but having a dynamic layout (specify each module to use 1/2 columns and adapt while placing) is no easy task, the user can reorder the modules so there are quite a lot of placements to test.
A configurable fixed number of columns...that may be easier...

**Comment by TrianguloY** [comment]
Yup, as I said the elements size is just "wrap content" so android decides. It's the easier, but has it's gotchas.

**Comment by Khoding** [comment]
Hello there, I'm a relatively new user (only been using for around three month) and I love the app, really changed the way I use my phone.
I was about to make a Feature Request to discuss basically the exact opposite. I don't mind it becoming bigger (on my phone anyway, I don't use my Tablet enough for it to bother me either way), I actually wish disabling incognito mode and merging Share and Copy buttons wouldn't make the app thinner on my Pixel 8 Pro.

![Image](https://github.com/user-attachments/assets/4e873c63-4cd2-4e51-8b53-82af717f7c26)
![Image](https://github.com/user-attachments/assets/ad99a93d-676f-40f0-9dd6-023826b32c5f)

I like the full width of screen it has when they're both activated, but then it becomes thinner and it looks odd being so thin.

I don't know what the solution is, but I was hoping we could maybe have an "Appearance" set of settings? that allows you to manually set width of the entire app?

My other idea, probably not ideal though, would be the ability to add "empty blocks" in the Open and Share module that allow you to keep the app the same size as when it has all the buttons or when they're all disabled.

**Comment by TrianguloY** [comment]
The current style is basically left as default, which makes it adapt to the screen elements. Less elements, less space. Maybe the style can be configured to have a fixed width instead? Does someone know of a way to improve the layout using plain android? (No external libraries, no appCompat, no androidx, and definitely no compose)

**Comment by Khoding** [comment]
So first of all, I'm not sure a global fixed width is the answer, it should be a user set per device fixed width, otherwise it will look wonky on a lot of devices, and the default should be how it is now (the width of the elements on screen).

Now to answer your question anyway, I haven't done that much android development, and when I did, it was mostly AndroidX, so I won't lie, I asked Claude AI to help me, what he gave me makes sense to me (it's what I remember learning at school 5 years ago anyway lmao, I also did a quick Google search just to make sure it gave me stuff that weren't complete nonsense).

```xml
<LinearLayout
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:orientation="vertical">

    <LinearLayout
        android:layout_width="300dp"  <!-- Fixed width in dp -->
        android:layout_height="wrap_content"
        android:layout_gravity="center_horizontal"
        android:orientation="vertical">
        
        <!-- Your content here -->
        
    </LinearLayout>
</LinearLayout>
```

and this as an alternative he produced, this could also be used to let the user define the width as I've stated. I'd go for this (or a similar concept) myself

```java
public class FixedWidthLayout extends LinearLayout {
    private int mFixedWidth;
    
    public FixedWidthLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
        // Initialize with default or from attrs
        mFixedWidth = 300; // dp value converted to pixels
    }
    
    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int fixedWidthSpec = MeasureSpec.makeMeasureSpec(
            mFixedWidth, MeasureSpec.EXACTLY);
        super.onMeasure(fixedWidthSpec, heightMeasureSpec);
    }
    
    public void setFixedWidth(int width) {
        mFixedWidth = width;
        requestLayout();
    }
}
```

As a third and potentially additional (I say additional because I think both of them are viable together, the user can either choose a width for the app, or a margin, whichever makes more sense to them, the set-width would be the same whatever the size of their screen, while the margin would fill in the screen with set margins on the side) solution, why not let the user choose the margin they want, as in, you let them set a inline-size by letting them choose their margin-inline (if they choose 32dp, it centers the app on screen with a margin from the side of their viewport of 32dp on both sides, and the app fills the in-between. For the elements inside the app, maybe give them a (okay sorry this is a CSS term but) [flex justify content of space-evenly](https://developer.mozilla.org/en-US/docs/Web/CSS/justify-content)?

I don't know if this helps, I'm just sharing my thoughts because we never know, might help you find the right solution for your app, even if they're stupid ideas lol.

**Comment by TrianguloY** [comment]
Thank you for the interest and samples.
I played with it and in the end I decided to add a setting to configure the dialog width. The current options are: dynamic (as it is now), full (will always fill the width, except padding) and any percentage between 1 and 99. I would have preferred to have a maxWidth or similar, and maybe someone will complain about not being able to configure a different value for portrait and landscape 😅 but for now I think this could help for your tablet layouts. I'll keep this issue open just in case. Don't hesitate to tell me what you think!

(note: if you want to test it you can install the evaluation [apk](https://github.com/TrianguloY/URLCheck/pull/434#issuecomment-2707392264) from the translations pr)

**Comment by Khoding** [comment]
Thanks, just tried it! I guess it does work very easily and as it doesn't seem to be a problem for most people, that should be enough. Thanks for the quick implementation 😄 do we know when the release with this feature will come out?

By the way I'll keep sharing the app with my friends because this really helped me with the mess that it opening links on Android so to me it's a 10/10 anyway. I don't know how your life is right now, but whenever I have paid work again, I'll try to support you so that you never stop working on it.
